src/coreclr/src/jit/morph.cpp

```diff
@@ -8652,58 +8652,68 @@ GenTree* Compiler::fgMorphOneAsgBlockOp(GenTree* tree)
         return nullptr;
     }
 
-    //
-    //  See if we can do a simple transformation:
-    //
-    //          GT_ASG <TYP_size>
-    //          /   \.
-    //      GT_IND GT_IND or CNS_INT
-    //         |      |
-    //       [dest] [src]
-    //
+    if ((destVarDsc != nullptr) && varTypeIsStruct(destLclVarTree) && destVarDsc->lvPromoted)
+    {
+        // Let fgMorphCopyBlock handle it.
+        return nullptr;
+    }
 
-    if (asgType == TYP_STRUCT)
+    if ((destVarDsc != nullptr) && !varTypeIsStruct(destVarDsc->TypeGet()))
     {
-        if (size == REGSIZE_BYTES)
+
+        //
+        //  See if we can do a simple transformation:
+        //
+        //          GT_ASG <TYP_size>
+        //          /   \.
+        //      GT_IND GT_IND or CNS_INT
+        //         |      |
+        //       [dest] [src]
+        //
+
+        if (asgType == TYP_STRUCT)
         {
-            if (clsHnd == NO_CLASS_HANDLE)
+            // It is possible to use `initobj` to init a primitive type on the stack,
```

Comment:
@kunalspathak thanks for your feedback, I added an example where `var` is not a struct, but `ASG` is.